### PR TITLE
issue fixed #8258 - assign indices for all array inputs so that validation works properly

### LINKED ISSUE #8258: M2.1.3 : Form validation with identical field names does not work as expected.
When you have multiple fields that share the same name, the JavaScript form validation does not work.

### Preconditions

1. Magento 2.1.3

### Steps to reproduce

Create a form with identical fields and add JavaScript validation. For example:

```
<form id="form-validate">
    <input type="text" name="street[]" class="required-entry" />
    <input type="text" name="street[]" class="required-entry" />
    <input type="text" name="street[]" class="required-entry" />
    <input type="submit" />
</form>
<script type="text/x-magento-init">
    {
        "#form-validate": {
            "validation": {}
        }
    }
</script>
```

### Expected result

When trying to submit the form, all 3 fields should show an error message according to their validation rule(s).

### Actual result

The message is only shown for the first field. Furthermore, the other fields are not validated at all.

### PR DESCRIPTION
**Pull request by jayankaghosh**
Form validation with identical field names does not work as expected

### Description
Assign indices to array inputs and later remove them before submitting as a workaround for the validation library to work properly.

### Fixed Issues (if relevant)
1. magento/magento2#8258 Form validation with identical field names does not work as expected

### Manual testing scenarios
1. Create a form with identical fields and add JavaScript validation
2. Submit the form with improper values in fields.
3. Validation should work properly now

### CODE REVIEW

File: lib/web/mage/validation.js [hunk 1]
```diff
@@ -1877,6 +1878,27 @@
             this._listenFormValidate();
         },
```
Comment (gautamkrishnar) [inline lib/web/mage/validation.js:1917]: Please add function documentation comment.

**Review by gautamkrishnar — CHANGES_REQUESTED** [review]

File: lib/web/mage/validation.js [hunk 2]
```diff
@@ -1877,6 +1878,27 @@
...
 
+        _workaroundForArrayInputs: function () {
```
Comment (VladimirZaets) [inline lib/web/mage/validation.js]: Please rename the method to something like "prepareArrayInputs" and add the `@see` annotation to doc block with link to issue (https://github.com/jquery-validation/jquery-validation/pull/717)

File: lib/web/mage/validation.js [hunk 3]
```diff
@@ -1877,6 +1878,27 @@
...
+            /* Store original names for array inputs */
+            var originalElements = [];
+
+            /* For all array inputs, assign index so that validation is proper */
+            this.element.find('[name$="[]"]').each(function (key, input) {
+                input = $(input);
+                var originalName = input.attr('name');
```
Comment (VladimirZaets) [inline lib/web/mage/validation.js]: Please move all var declaration to the top of the function.

**Comment by VladimirZaets** [comment]
Hi @jayankaghosh  , thanks for fixes.

**Review by gautamkrishnar — APPROVED** [review]
Awesome

**Comment by jayankaghosh** [comment]
Oh didn't follow on this for a long time. So this is fixed now? Cool!
P.S I can see the Travis CI failed due to code standard not being proper. So does that need to be reworked on?

**Comment by VladimirZaets** [comment]
Hi @jayankaghosh. Thanks for collaboration.

After the testing we found the next problem:

This fix doesn't work with checkbox input, for instance with downloadable product.
Manual failing scenario:
1. Create downloadable product with two links and enabled "Links can be purchased separately" checkbox.
2. Navigate to storefront and try to add created product to cart with only one link checked on PDP.

Expected result:
Product added successfully to cart with only one link selected.

Actual result: 
Validation failed with message: "Please select one of the options."

**Comment by jayankaghosh** [comment]
> Hi @jayankaghosh. Thanks for collaboration.
> 
> After the testing we found the next problem:
> 
> This fix doesn't work with checkbox input, for instance with downloadable product.
> Manual failing scenario:
> 
> 1. Create downloadable product with two links and enabled "Links can be purchased separately" checkbox.
> 2. Navigate to storefront and try to add created product to cart with only one link checked on PDP.
> 
> Expected result:
> Product added successfully to cart with only one link selected.
> 
> Actual result:
> Validation failed with message: "Please select one of the options."

I am working on this

**Comment by jayankaghosh** [comment]
@magento-engcom-team give me magento2.3 instance

**Comment by jayankaghosh** [comment]
@magento-engcom-team give me 2.3.0 instance

**Comment by magento-engcom-team** [comment]
Hi @jayankaghosh. Thank you for your request. I'm working on Magento 2.3.0 instance for you

**Comment by magento-engcom-team** [comment]
Hi @jayankaghosh, here is your [Magento instance](https://i-15383-2-3-0.instances.magento-community.engineering).
Admin access: [https://i-15383-2-3-0.instances.magento-community.engineering/admin](https://i-15383-2-3-0.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`
Instance will be terminated in up to 3 hours.

**Comment by jayankaghosh** [comment]
@VladimirZaets I couldn't reproduce the issue with the steps mentioned

**Comment by sdzhepa** [comment]
@magento-engcom-team give me test instance

**Comment by magento-engcom-team** [comment]
Hi @sdzhepa. Thank you for your request. I'm working on Magento instance for you

**Comment by magento-engcom-team** [comment]
Hi @sdzhepa, here is your [new Magento instance](https://pr-15383.instances.magento-community.engineering).
Admin access: [https://pr-15383.instances.magento-community.engineering/admin](https://pr-15383.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`

**Comment by magento-engcom-team** [comment]
Hi @sdzhepa, here is your [Magento instance](https://i-15383-2-3-0.instances.magento-community.engineering).
Admin access: [https://i-15383-2-3-0.instances.magento-community.engineering/admin](https://i-15383-2-3-0.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`
Instance will be terminated in up to 3 hours.

**Comment by sdzhepa** [comment]
@magento-engcom-team give me 2.3-develop instance

**Comment by magento-engcom-team** [comment]
Hi @sdzhepa. Thank you for your request. I'm working on Magento 2.3-develop instance for you

**Comment by magento-engcom-team** [comment]
Hi @sdzhepa, here is your [Magento instance](https://i-15383-2-3-develop.instances.magento-community.engineering).
Admin access: [https://i-15383-2-3-develop.instances.magento-community.engineering/admin](https://i-15383-2-3-develop.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`
Instance will be terminated in up to 3 hours.

**Comment by sdzhepa** [comment]
Hello @jayankaghosh 

I have double checked scenario described here https://github.com/magento/magento2/pull/15383#issuecomment-458337123 

And can confirm:
1. scenario works on `2.3-develop` branch now. Product is added to cart with only one link
2. scenario does not work on branch with this PR.  Product is NOT added to cart with only one link

I also see in https://github.com/magento/magento2/pull/15383#issuecomment-458486719 that you requested 
> give me 2.3.0 instance

And it could be a reason why it is not reproduced.  More details [Deploy instance based on PR changes](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#deploy-pr)

> To verify and test changes completed in a pull request..
> (at)magento-engcom-team give me test instance

Could you please confirm and fix found issue?


Thank you for contribution and collaboration!

**Comment by magento-engcom-team** [comment]
Hi @sdzhepa. Thank you for your request. I'm working on Magento instance for you

**Comment by magento-engcom-team** [comment]
Hi @sdzhepa, here is your [new Magento instance](https://pr-15383.instances.magento-community.engineering).
Admin access: [https://pr-15383.instances.magento-community.engineering/admin](https://pr-15383.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`

**Comment by jayankaghosh** [comment]
@sdzhepa I have identified the problem. Will think of a solution and push when I get time.

**Comment by jayankaghosh** [comment]
@magento-engcom-team  give me test instance

**Comment by VladimirZaets** [comment]
Hi @jayankaghosh. We will retest tour PR and leave feedback. Thanks.

**Comment by p-bystritsky** [comment]
@jayankaghosh, thank you for contribution.
Unfortunately, we have faced an issue. You can reproduce it by the next steps:
1. Create a Product.
2. Create a Cart Price Rule with some Discount Code.
3. Go to the Storefront.
4. Add the Product to Cart.
5. Edit Cart.
6. Try to apply the Discount Code.
An error message appears: 
Screenshot_5

**Comment by jayankaghosh** [comment]
@magento-engcom-team give me test instance

**Comment by magento-engcom-team** [comment]
Hi @jayankaghosh. Thank you for your request. I'm working on Magento instance for you

**Comment by magento-engcom-team** [comment]
Hi @jayankaghosh, here is your [new Magento instance](https://pr-15383.instances.magento-community.engineering).
Admin access: [https://pr-15383.instances.magento-community.engineering/admin](https://pr-15383.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`

**Comment by jayankaghosh** [comment]
@p-bystritsky I couldn't reproduce the issue. Could you please have a look again and/or mention a more detailed steps to reproduce

**Comment by p-bystritsky** [comment]
@jayankaghosh, I have a clean installation of Magento 2 with preconditions, visible below:
invalid_form_key

**Comment by jayankaghosh** [comment]
@p-bystritsky Thanks for informing. I'll have a look

**Comment by VladimirZaets** [comment]
Hi @jayankaghosh, I am closing this PR now due to inactivity.
Please reopen and update if you wish to continue.
Thank you for collaboration

**Comment by jayankaghosh** [comment]
@magento-engcom-team give me test instance

**Comment by magento-engcom-team** [comment]
Hi @jayankaghosh. Thank you for your request. I'm working on Magento instance for you

**Comment by magento-engcom-team** [comment]
Hi @jayankaghosh, here is your [new Magento instance](https://pr-15383.instances.magento-community.engineering).
Admin access: [https://pr-15383.instances.magento-community.engineering/admin](https://pr-15383.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`

**Review by VladimirZaets — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @VladimirZaets, thank you for the review.
[ENGCOM-5174](https://jira.corp.magento.com/browse/ENGCOM-5174) has been created to process this Pull Request

**Comment by VasylShvorak** [comment]
Hi @jayankaghosh Unfortunately PR changes cause new issues.
**Case 1:**
- Create a Product.
- Create a Cart Price Rule with some Discount Code.
- Go to the Storefront.
- Add the Product to Cart.
- Edit Cart.
- Enter discount code
- Click button Apply Discount
**Result:** 
:x: Nothing happen. Discount cannot be applied from cart page.

**Case 2:**
- Go to Create New Customer Account page or Customer Login page
- Fiil all fields
- Try login or create account
**Result:**
:x: Buttons Create an Account and Sign In are not working. Customer cannot be created or logged in.

**Comment by jayankaghosh** [comment]
@VasylShvorak This issue is resolved now. 
The only drawback is, if the `mage/validation` plugin is used on a form, no other form submit listeners should be attached to the form, or, a `submitHandler` has to be explicitly mentioned where the further form submit listeners need to be specified. 
I believe this is a very particular and rare scenario and should not happen that often. It should work with all existing Magento's forms which are using the `mage/validation`

**Review by VladimirZaets — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @VladimirZaets, thank you for the review.
[ENGCOM-5174](https://jira.corp.magento.com/browse/ENGCOM-5174) has been created to process this Pull Request

**Comment by VasylShvorak** [comment]
:heavy_check_mark: QA passed

**Comment by magento-cicd2** [comment]
[CLA assistant check](https://cla.dev.magento.com/magento/magento2?pullRequest=15383) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla.dev.magento.com/magento/magento2?pullRequest=15383) before we can accept your contribution.<br/>**3** out of **4** committers have signed the CLA.<br/><br/>:white_check_mark: jayankaghosh<br/>:white_check_mark: VladimirZaets<br/>:white_check_mark: p-bystritsky<br/>:x: ysapiga

**Comment by magento-cicd2** [comment]
[CLA assistant check](https://cla.dev.magento.com/magento/magento2?pullRequest=15383) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla.dev.magento.com/magento/magento2?pullRequest=15383) before we can accept your contribution.<br/>**3** out of **4** committers have signed the CLA.<br/><br/>:white_check_mark: jayankaghosh<br/>:white_check_mark: VladimirZaets<br/>:white_check_mark: p-bystritsky<br/>:x: ysapiga

**Pull request merged**